torch/_dynamo/variables/builtin.py

```diff
@@ -1677,11 +1677,16 @@ def _unimplemented():
         ):
             return ConstantVariable.create(op(left.value, right.value))
 
-        if (
-            (isinstance(left, StreamVariable) and isinstance(right, StreamVariable))
-            or (isinstance(left, EventVariable) and isinstance(right, EventVariable))
-        ) and op is operator.eq:
-            return ConstantVariable(op(left.value, right.value))
+        if isinstance(left, (StreamVariable, EventVariable)) or isinstance(
+            right, (StreamVariable, EventVariable)
+        ):
+            if type(left) == type(right) and op is operator.eq:
+                return ConstantVariable(op(left.value, right.value))
```

Comment:
Are we guarding on this?  I think this might be incorrect.

Can you add a test like:
```
def fn(a, s0, s1):
	if s0 == s1:
   		return a + 1
	else
		return a - 1
```

Then
1) First compile it where s0 != s1
2) Then run it again with s0 == s1 (to ensure we recompile due to the guard)

And also try the opposite of that.

Reply:
We have a ID_MATCH guard when creating the StreamVariable. But let me add the test to make sure.

Reply:
The added test passes. Had a discussion with @yanboliang.

ID_MATCH may not be an accurate guard. It may be too strict because we might want to re-use the compiled artifact for one stream when running in the other stream. If that's the case, we don't want to perform an ID_MATCH. But I remember the discussion, where we don't want to preserve computation stream's information since inductor will re-allocate them but not so sure about communication streams.

Reply:
Without an ID_MATCH we would need to guard on `s0!=s1`.